docs: Create CODE_OF_CONDUCT.md (#28633)
renovatebot/renovate: CODE_OF_CONDUCT.md

We want to avoid maintainers getting stressed out by bad behavior from contributors.
That's why we have these rules.

## Politeness

Sadly, it's common in Open Source projects for a few users to behave in an aggressive and rude way.
A user might say something like: "You should have fixed this bug already!", or "Why am I still waiting for this feature?".
We do not allow this kind of behavior.

We expect basic politeness, do not act rude.
For example: it is okay if you ask a question and do not thank us afterwards.
But avoid writing mean comments like: "Pity the documentation didn’t say that." or "Thanks for nothing.".

## Respect the time of those who help you

Respect goes both ways, but time is limited.
When you ask for help, please remember that the maintainer's time is valuable.
We get many questions each week and do our best to answer each one.
To get the help you need, please be prepared to give detailed logs or descriptions of your issues.
If you do not want to spend the effort giving us enough information, it's likely you will not get the help you need.

Remember, most of the support provided by our team, including the Mend.io staff, is _unpaid_.

> ## Blocking and unblocking

Reviewer: I found a section about blocking in the [`contributing.md` file on your branch](https://github.com/renovatebot/renovate/blob/docs/coc/.github/contributing.md):

> ## Rate Limiting of Support Requests through Temporary Blocking
> 
> To ensure that the Renovate maintainers don't burn out from dealing with unfriendly behavior, those who display a bad attitude when asking for or receiving support in the repo will be rate limited from further requests through the use of temporary blocking.
> The duration of the temporary block depends on how rude or inconsiderate the behavior is perceived to be, and can be from 1-30 days.
> 
> If you have been blocked temporarily and believe that it is due to a misunderstanding, or you regret your comments and wish to make amends, please reach out to the lead maintainer Rhys Arkins by email with any request for early unblocking.
> If/once you are unblocked, you should edit or delete whatever comment lead to the blocking, even if you did not intend it to be rude or inconsiderate.
> Long emails or apologies are undesirable - the maintainers are busy and want to be able to help as many users as possible with the time they have available.

I think the parts that are still relevant should be edited and moved to the Code of Conduct. Then add a link to the Code of Conduct in place of the old text.

It's probably good to mention how to get "unbanned/unblocked" in the Code of Conduct. And how long/short you want the apology to be.

We quickly deal with rudeness in the community with:

- automated comments
- temporary blocks
- permanent bans

If you keep breaking the rules or challenge our guidelines openly, you will be blocked.
For example: if you keep spamming the maintainers with `@mentions` or challenge our rules openly, you will be blocked.

We generally do not argue about these decisions, but we are willing to reverse a block if a you show that you understand and respect the rules, or if there was a misunderstanding.

Simply put: we block and unblock swiftly, what matters is how you follow the rules going forward.

## How we prioritize work

Renovate's core contributors and maintainers focus on work that:

- Helps a lot of users, or
- Fixes regressions (errors introduced by recent changes), or
- Is required by a customer of Mend.io, or
- Is sponsored by third parties after independent validation, or
- We personally need or want to implement

You may be disappointed when we focus on other work ahead of your feature or bug, but you should understand and accept this.

> ## Maintaining Issue and Code Quality

Reviewer: ```suggestion
## Maintaining Issue and Code quality
```

This way most issues are ready to work on, either by us or the community.

We may reject ideas that are too specialized, or that would make the project too hard to maintain.

We have strict coding standards and reviews to keep our code in good shape.
A feature or fix must of course work, but it must also be well designed to stay maintainable.
We may ask you to improve your code several times in a row, which can be difficult for you.
We only do this to keep the project sustainable.

## If you have urgent work

People working for big companies might push too hard in Open Source projects.
It’s often hard for them to understand that our maintainers cannot spend much time to solve their issues quickly.
Frequent requests for updates like "@rarkins how can we move this forward?" are _not_ helpful.

Please remember, unless you are a Mend.io customer, this project does not owe you the level of response or support you might expect.
Mend.io customers should use their designated support channels for urgent needs.

## Getting more help

If you need more assistance than what this project offers, you have two options:

1. Become a Mend.io customer, such as by buying Renovate Enterprise, or

> 1. Hire an experienced Renovate contributor privately for consulting. Mend.io staff do _not_ offer this service, but one of our volunteer maintainers, `@secustor`, does

Reviewer: ```suggestion
1. Hire an experienced Renovate contributor privately for consulting. Mend.io staff do _not_ offer this service, but one of our volunteer maintainers, but [`@secustor`](https://github.com/secustor), does
```

Might be good to link to their profile page here.

> 1. Hire an experienced Renovate contributor privately for consulting. Mend.io staff do _not_ offer this service, but one of our volunteer maintainers, but [`@secustor`](https://github.com/secustor), does

Reviewer 2: double `but`?

Reviewer: ```suggestion
1. Hire an experienced Renovate contributor privately for consulting. Mend.io staff do _not_ offer this service, but one of our volunteer maintainers, [`@secustor`](https://github.com/secustor), does
```

You're right, good catch! :heart:

## Feedback

> We welcome respectful discussions about this policy and accept suggestions that improve this text.

Reviewer: ```suggestion
We welcome respectful discussions about these rules and accept suggestions that improve this text.
```

# Code of Conduct

In order to deliver great features and support to the Renovate Open Source project, we request a high level of politeness and diligence from those wishing to participate here.

## Our priorities

We prioritize the sustainability of this project, which means prioritizing maintainers and contributors, many of whom donate their time to helping others.
The number one risk to Open Source projects is maintainer burnout, and the number one cause of maintainer burnout is impoliteness in various forms, usually from a minority of users.

## Politeness

It is unfortunately common in Open Source projects for a minority of users act with aggression and disrespect (e.g. "How dare you not fix this already"), but we do not accept it here.

> We don't expect too much in terms of overt politeness - simply don't be impolite.

Reviewer 2: ```suggestion
We don't expect too much in terms of over politeness - simply don't be impolite.
```

Author: Overt was intentional

It's also important to understand that the majority of time you see spent supporting users is voluntary and not paid - even by Mend.io staff.

## Blocking and unblocking

We act quickly and decisively to address impoliteness from community members through automated comments, or temporary or permanent banning.
If you display repeated undesirable behavior (e.g. spamming maintainers with @mentions) or state openly your defiance to the rules ("How dare you ask me not to @mention you?") then you'll receive a block.
We do not engage in "debate" about such blocks, but we'll almost always undo a block for any user who demonstrates a newfound understanding of the rules or explains it was a misunderstanding.

In other words: we can be quick to both block and unblock, and all that matters is that you can conduct yourself in future according to the project's rules.

## How we prioritize work

When it comes to Renovate's core contributors and maintainers, we prioritize features and fixes:

- those affecting a large number of users, or
- regression errors (something used to work, but we broke it by accident), or

> - features needed by a Mend.io customer (Renovate's sponsor/owner), or

Reviewer 3: ```suggestion
- features needed by a Mend.io customer (Renovate's sponsor/owner), or
- are sponsored by 3rd parties after independent validation ( PRs will not be accepted if they are not in the interest of Renovate OSS ), or
```
Formulating it like that so people don't think if they are ok with paying, we will implement anything.

- those we need personally, or just want to see done

A surprising number of users find it hard to accept that the Renovate core team or other community contributors won't build every feature or fix every bug which they'd like to see.

> Participating in this repo is dependent on you finding a way to accept this reality, and not being pushy about it.

Reviewer 3: This sounds pretty aggressive to me.

We may apply multiple rounds of code review and suggestions, and sometimes this will feel like a burden, but ultimately our goal is to maintain a sustainable and maintainable project and this is a higher priority than whatever feature or fix you are trying to get merged.

## If you have important work to get done

People who work for large and/or important companies sometimes let the "pressure" of their role or task spill over into pushiness in Open Source.
Such people sometimes find it hard to accept that maintainers cannot spend significant time helping them troubleshoot a config issue, or talk them through logs, or providing a feature or fix they need immediately.
They are especially prone to "pinging" a maintainer with comments like `@rarkins how can we move this forward?`.

Please understand that you are not our customer, and this Open Source project like any other does not owe you even the responsiveness or support you're probably already receiving.
If you do happen to be a Mend.io customer then escalation of feature or support requests should be done via your Support or Customer Success representative and not in Open Source.

## Getting more help

> If you need more help than this Open Source project can provide, or simply need guaranteed help, then there are two main options:

Reviewer 3: ```suggestion
If you need more help for you than this Open Source project can provide, simply need guaranteed help or need a feature which has not been prioritised high enough for you, then there are two main options:
```

1. Become a customer of Mend.io, such as purchasing Renovate Enterprise, or

> 2. Engage privately with an experienced Renovate contributor on an hourly or project basis. Mend.io staff do not currently offer this, but one of Renovate's non-Mend.io employed maintainers who otherwise works voluntarily on the project is available for such consulting work.

Reviewer 3: ```suggestion
2. Engage privately with an experienced Renovate contributor on an hourly or project basis. Mend.io staff do not currently offer this, but one of Renovate's non-Mend.io employed maintainers ( @secustor ) who otherwise works voluntarily on the project is available for such consulting work.
```

# Code of Conduct

> In order to deliver great features and support to the Renovate Open Source project, we ask for high politeness and careful attention from those who want to join us.

Author: ```suggestion
In order to deliver great features and support to the Renovate Open Source project, we ask for high politeness and careful attention from those who want to participate.
```

> In order to deliver great features and support to the Renovate Open Source project, we ask for high politeness and careful attention from those who want to participate.

Reviewer: ```suggestion
To help us deliver great features and support the Renovate Open Source project we ask that you:

- are polite
- pay attention to details
- keep in mind that most maintainers are volunteers
- are respectful of the time and effort of the maintainers
```

Most people won't read the whole thing. So we should try to put the main points up front. I think this covers most of the big ticket items.

## Our priorities

> Our main goal is to keep this project sustainable.
> This involves supporting our maintainers and contributors, most of whom volunteer their time to help others.
> The biggest threat to Open Source projects is when maintainers get too stressed, and this stress is often caused by rudeness from a few users.

Reviewer: ```suggestion
We want to keep this project sustainable.
This means we support our maintainers and contributors, who spend their free time to help others.

Maintainers getting stressed is a big threat to Open Source projects, like ours.
Stressed maintainers quit, or reduce their time spent on the project.
Often a few users behave badly, where most users are nice.

We want to avoid maintainers getting stressed out by bad behavior from contributors.
That's why we have these rules.
```

## Politeness

> Sadly, it's not unusual in Open Source projects for a few users to be aggressive and disrespectful (like saying, "You should have fixed this already"), but we do not tolerate this behavior here.

Reviewer: ```suggestion
Sadly, it's common in Open Source projects for a few users to behave in a aggressive and rude way.
A user might say something like: "You should have fixed this bug already!", or "Why am I still waiting for this feature?".
We do not allow this kind of behavior.
```

> We expect basic politeness—just don’t be rude.

Reviewer: ```suggestion
We expect basic politeness, do not act rude.
```

> It's fine if you ask a question and don’t thank us afterward.

Reviewer: ```suggestion
For example: it is okay if you ask a question and do not thank us afterwards.
```

> However, avoid making disrespectful remarks like "Pity the documentation didn’t say that" or "thanks for nothing."

Reviewer: ```suggestion
But avoid writing mean comments like: "Pity the documentation didn’t say that." or "Thanks for nothing.".
```

## Respect the time of those who help you

> Respect is a two-way street, but time is limited, and those asking for help should remember that our maintainers’ time is precious.

Reviewer: ```suggestion
Respect goes both ways, but time is limited.
When you ask for help, please remember that the maintainer's time is valuable.
```

> We receive many queries each week and do our best to address each one.
> Often, we need you to provide detailed logs or descriptions of your issues.
> If you’re not prepared to do this, it’s unlikely you will get the help you need.

Reviewer: ```suggestion
We get many questions each week and do our best to answer each one.
To get the help you need, please be prepared to give detailed logs or descriptions of your issues.
If you do not want to spend the effort giving us enough information, it's likely you will not get the help you need.
```

> Remember, most of the support provided by our team, including Mend.io staff, is unpaid.

Reviewer: ```suggestion
Remember, most of the support provided by our team, including the Mend.io staff, is _unpaid_.
```

## Blocking and unblocking

> We quickly tackle rudeness in the community through automated comments, temporary blocks, or permanent bans.
> If you repeatedly break the rules (like spamming maintainers with @mentions) or challenge our guidelines openly, you will be blocked.
> We generally do not argue about these decisions, but we are willing to reverse a block if a user shows they understand and respect the rules or if there was a misunderstanding.
> 
> Simply put, we block and unblock swiftly, and what matters is how you follow the rules going forward.

Reviewer: ```suggestion
We quickly deal with rudeness in the community with:

- automated comments
- temporary blocks
- permanent bans

If you keep breaking the rules or challenge our guidelines openly, you will be blocked.
For example: if you keep spamming the maintainers with `@mentions` or challenge our rules openly, you will be blocked.

We generally do not argue about these decisions, but we are willing to reverse a block if a you show that you understand and respect the rules, or if there was a misunderstanding.

Simply put: we block and unblock swiftly, what matters is how you follow the rules going forward.
```

## How we prioritize work

> For Renovate's core contributors and maintainers, we focus on work that:
> 
> - Benefits a large number of users, or
> - Fixes regressions (errors introduced by recent changes), or
> - Is required by a Mend.io customer,
> - Is sponsored by 3rd parties after independent validation, or
> - We personally need or want to implement
> 
> It may be disappointing when we can’t prioritize your specific feature or bug, but understanding and accepting this is crucial.

Reviewer: ```suggestion
Renovate's core contributors and maintainers focus on work that:

- Helps a lot of users, or
- Fixes regressions (errors introduced by recent changes), or
- Is required by a customer of Mend.io, or
- Is sponsored by third parties after independent validation, or
- We personally need or want to implement

You may be disappointed when we focus on other work ahead of your feature or bug, but you should understand and accept this.
```

## Maintaining Issue and Code Quality

> We use GitHub Discussions to start and sort issues but don't allow non-maintainers to directly create new issues.

Reviewer: ```suggestion
We use GitHub Discussions to start and sort issues.
Only maintainers are allowed to create new issues.
```

If we confirm a bug or agree with a feature idea, and if it's well-documented, we will turn it into an official issue.

> This process ensures that most issues are actionable, either by us or the community.

Reviewer: ```suggestion
This way most issues are ready to work on, either by us or the community.
```

> Sometimes we reject ideas if they are too specialized and would complicate the project more than they would help.

Reviewer: ```suggestion
We may reject ideas that are too specialized, or that would make the project too hard to maintain.
```

> We enforce strict coding standards and reviews to keep our codebase maintainable.

Reviewer: ```suggestion
We have strict coding standards and reviews to keep our code in good shape.
```

> A feature or fix must not only work, it must be well designed to stay maintainable.

Reviewer: ```suggestion
A feature or fix must of course work, but it must also be well designed to stay maintainable.
```

> We may ask for several revisions, which can be demanding, but our goal is to keep the project sustainable.

Reviewer: ```suggestion
We may ask you to improve your code several times in a row, which can be difficult for you.
We only do this to keep the project sustainable.
```

> A surprising number of users find it hard to accept that the Renovate core team or other community contributors won't build every feature or fix every bug which they'd like to see.
> Participating in this repo is dependent on you finding a way to accept this reality, and not being pushy about it.

Author: ```suggestion
When users are excited or pressured to get an implementing working then it can be hard to accept that there's nobody to add their particular feature or look into their bug, but it's important that you understand and accept this reality.
```

## If you have urgent work

> People working for significant companies might push too hard in Open Source projects.
> It’s often hard for them to understand that our maintainers cannot dedicate much time to solve their specific issues quickly.
> Frequent requests for updates like "@rarkins how can we move this forward?" are not helpful.

Reviewer: ```suggestion
People working for big companies might push too hard in Open Source projects.
It’s often hard for them to understand that our maintainers cannot spend much time to solve their issues quickly.
Frequent requests for updates like "@rarkins how can we move this forward?" are _not_ helpful.
```

Please remember, unless you are a Mend.io customer, this project does not owe you the level of response or support you might expect.
Mend.io customers should use their designated support channels for urgent needs.

## Getting more help

> If you need more assistance than what this project offers, there are two options:

Reviewer: ```suggestion
If you need more assistance than what this project offers, you have two options:
```

> 1. Become a Mend.io customer, such as by purchasing Renovate Enterprise, or

Reviewer: ```suggestion
1. Become a Mend.io customer, such as by buying Renovate Enterprise, or
```

> 2. Hire an experienced Renovate contributor privately for consulting. Mend.io staff do not offer this service, but one of our volunteer maintainers, `@secustor`, does.

Reviewer: ```suggestion
1. Hire an experienced Renovate contributor privately for consulting. Mend.io staff do _not_ offer this service, but one of our volunteer maintainers, `@secustor`, does
```

## Feedback

> We welcome respectful discussions about this policy and accept suggestions for improving this text.

Reviewer: ```suggestion
We welcome respectful discussions about this policy and accept suggestions that improve this text.
```

We avoid debates on social media or going off-topic in GitHub Discussions.

> Our consistent delivery of new features and excellent community support show that our approach is effective.

Reviewer: ```suggestion
Because we enforce all these rules, we can deliver new features and give excellent support to the community.
```
